### ISSUE TITLE
[MKL-DNN] Failure to run face model (demark)

### ISSUE THREAD

**Issue by jczaja** [PaddlePaddle/Paddle#18658]
Hi,

We have just tested that internal face model of yours (demark) stopped to work with mkl-dnn (it is fine when paddle naive ops are used). 

commit tested:
fd6631ef2fb8c896f89ddaa94a5112c08665f35b

commandline used:
`
./paddle/fluid/inference/tests/api/test_analyzer_image_classification  --infer_model=/home/jczaja/models/demark/demark/ --gtest_filter=*profile_mkldnn --batch_size=1
`
Output:
`
Exception: /home/jczaja/paddle/paddle/fluid/memory/detail/meta_cache.cc:33 Assertion `desc->check_guards()` failed.
terminate called after throwing an instance of 'std::runtime_error'
what():  Exception encounter.
Aborted
`

We do not know yet when exactly problem appeared 

@luotao1 Could you please confirm that demark model does fail  for you when mkl-dnn is used? If all works for you then tell us revision used and commandline

**Comment by luotao1** [comment]
I will confirm it ASAP.

**Comment by luotao1** [comment]
I could run successfully with the latest develop branch on demark model.
please check the md5sum for it.
```
e5fa00447f90969a95e2b77fd5b77c26  demark/model
9c8ab6ea9db3b5d59b70d68bf0086524  demark/params
```
command:
```
./test_analyzer_image_classification --gtest_filter=Analyzer_resnet50.profile_mkldnn --batch_size=1 --warmup --repeat=100 --paddle_num_threads=4 --infer_model=face_model/demark
```

**Comment by grygielski** [comment]
@luotao1 it keeps failing for me, md5sums are fine. Could you provide me build command you used?

**Comment by luotao1** [comment]
Which mkldnn version do you use? 
The build command is in https://github.com/PaddlePaddle/Paddle/issues/18658#issuecomment-513090602

**Comment by grygielski** [comment]
I meant cmake command you used to build PaddlePaddle. I just cloned current develop branch so mkldnn version is 0.18

**Comment by luotao1** [comment]
How about give your CMake command at first?

**Comment by luotao1** [comment]
My cmake command is 
```
cmake .. -DON_INFER=ON -DWITH_GPU=OFF -DWITH_MKLDNN=ON -DCMAKE_BUILD_TYPE=Release -DWITH_TESTING=ON -DWITH_PYTHON=ON -DWITH_INFERENCE_API_TEST=ON
```
@grygielski

**Comment by grygielski** [comment]
I've been using same command for building. We'll have to take look at it.

**Comment by jczaja** [comment]
Status: 
Problem root caused. Mkl-dnn activation op for getting y tensor allocated is using:
`T *y_data = y->mutable_data<T>(ctx.GetPlace());`
This is wrong, for out of place computation as when blocked formats (NCHW16C) are used 
allocation may be a bit bigger that N\*C\*H\*W\*sizeof(data_type), as implicit padding may be present.

Proper allocation should be:
` T *y_data = y->mutable_data<T>(ctx.GetPlace(), activation_pd->dst_primitive_desc().get_size());`

as destination primitive descriptor holds total number of allocation size.

#### Notes:
- If allocation is not big enough then it is likely that meta-data of next chunk of memory
in buddy allocator will be overwritten resulting in a crash as presented in first entry of this issue.
- Problem does not manifest on Face Model of BDW platform because activation (tanh) is using
NCHW format then no implict padding is used by mkl-dnn so size of allocation is proper

**Comment by luotao1** [comment]
> Problem does not manifest on Face Model of BDW platform

What's the meaning of BDW? I still don't know why I could run OK on this model but you got fails. What's the reason of it, different machine?

**Comment by grygielski** [comment]
I guess Jacek meant that You are running models on Broadwell (BDW) architecture: https://en.wikipedia.org/wiki/Broadwell_(microarchitecture). It doesn't support AVX-512 processor operations (https://en.wikipedia.org/wiki/Advanced_Vector_Extensions) so it uses older ones which have different data formats (e.g. NCHW instead of NCHW16C). @luotao1

**Comment by luotao1** [comment]
Got it. We use E5-2650 v4 CPU, which only support AVX2.

**Comment by jczaja** [comment]
Status of investigation: Failure (accuracy problems) are a result that fetch operator does not perform conversion of mkl-dnn data into paddle data. OperatorsWithKernel are invoking PrepareData which may trigger conversion from MKL-DNN data format into Paddle data format. For face model just before fetch op there is "tanh" operator which is executed by mkl-dnn . For AVX512 tanh will work on mkl-dnn format nchw16c  and will send tensor of such data format into fetch op. Fetch assumes this tensor is NCHW and will copy it to output. NCHW16C is diffrent/not compatible with NCHW so this is a problem. To fix this  conversion(reorder) has to be added before fetch op.

Fast workaround:
- Disable tanh op in face model eg. use paddlepaddle tanh activation. MKL-DNN tanh is not faster.
So no performance penalty is by disabling tanh.

@luotao1 , @Superjomn  Please advice how to proceed . Eg. Where to call TransDataLayoutFromMKLDNN  (data_layout_transform.cc:~119) to have output from MKL-DNN ops converted for the purpose of consuming it by fetch op ?

**Comment by luotao1** [comment]
@jczaja  Could the accuracy diff be solved If there is only first `PrepareData` in `OperatorsWithKernel`, not each time?

**Comment by luotao1** [comment]
> Disable tanh op in face model eg. use paddlepaddle tanh activation. MKL-DNN tanh is not faster.
So no performance penalty is by disabling tanh.

I think it's Ok for just this model on avx512. 

> Where to call TransDataLayoutFromMKLDNN (data_layout_transform.cc:~119) to have output from MKL-DNN ops converted for the purpose of consuming it by fetch op ?

@LeoZhao-Intel @jianhang-liu How do you think about it?

**Comment by LeoZhao-BJ** [comment]
> > Disable tanh op in face model eg. use paddlepaddle tanh activation. MKL-DNN tanh is not faster.
> > So no performance penalty is by disabling tanh.
> 
> I think it's Ok for just this model on avx512.
> 
> > Where to call TransDataLayoutFromMKLDNN (data_layout_transform.cc:~119) to have output from MKL-DNN ops converted for the purpose of consuming it by fetch op ?
> 
> @LeoZhao-Intel @jianhang-liu How do you think about it?

Just my 2 cents:  
Currently for ops based on OperatorsWithKernel, PrepareData is just doing data layout transform on input tensors instead of outputs, in one graph it follows this rule and works well.
But this case is special since last op based on OperatorsWithKernel is connected with fetch op which is not based on OperatorsWithKernel, then there is issue.
2 options I think, one is do layout conversion for output in OperatorsWithKernel, but I don't think it is good solution, since it don't know which format is needed by next op. The other is doing conversion in fetch op, which I think is good.

**Comment by jczaja** [comment]
Ok, I will work on adding conversion into fetch op eg.call TransDataLayoutFromMKLDNN inside fetch::RunImpl

**Comment by jczaja** [comment]
@LeoZhao-Intel , @luotao1 PR #19282 with my proposal of modifications to fetch op so it solves most recent issue discussed here (lack of conversion from MKL-DNN to Paddle in fetch op). It works for Face model and some other testing I have done. Please take a look and let me know what you think.

**Comment by jczaja** [comment]
Fixed in #19282
